File: scraper/scrape_foodpanda.py
#!/usr/bin/env python3
"""
Foodpanda 香港健康餐爬蟲 (Python 版)
使用方式: pip install playwright && playwright install chromium && python3 scrape_foodpanda.py
"""

import json
import time
from pathlib import Path

try:
    from playwright.sync_api import sync_playwright
except ImportError:
    print("請先安裝: pip install playwright")
    print("然後執行: playwright install chromium")
    exit(1)

OUTPUT = Path(__file__).parent.parent / "scraped-foodpanda.json"
URL = "https://www.foodpanda.hk/city/hong-kong/cuisine/healthy"


def scrape():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        page = browser.new_page()
        page.set_viewport_size({"width": 1280, "height": 800})
        page.set_extra_http_headers({"Accept-Language": "zh-HK,zh;q=0.9"})

        print("前往 Foodpanda 健康餐分類...")
        page.goto(URL, wait_until="networkidle", timeout=30000)
        time.sleep(2)

        # 滾動載入
        page.evaluate("""
            async () => {
                await new Promise(r => {
                    let h = 0;
                    const t = setInterval(() => {
                        window.scrollBy(0, 300);
                        h += 300;
                        if (h >= document.body.scrollHeight) { clearInterval(t); r(); }
                    }, 100);
                });
            }
        """)
        time.sleep(2)

        # 除錯：儲存頁面 HTML 以便檢查結構
        debug_html = Path(__file__).parent.parent / "debug-foodpanda.html"
        debug_html.write_text(page.content(), encoding="utf-8")
        print(f"頁面 HTML 已儲存至 {debug_html}，可檢查 DOM 結構")

        restaurants = page.evaluate("""
            () => {
                const results = [];
                const cards = document.querySelectorAll('a[href*="/restaurant/"]') ||
                    document.querySelectorAll('[data-testid="restaurant-card"]') ||
                    document.querySelectorAll('[class*="RestaurantCard"]') ||
                    document.querySelectorAll('article');
                cards.forEach((card, i) => {
                    const link = card.closest('a') || card.querySelector('a[href*="/restaurant/"]') || card;
                    const href = link.href || '';
                    const nameEl = card.querySelector('h2, h3, [class*="name"], [class*="title"]');
                    const name = nameEl ? nameEl.textContent.trim() : '';
                    const addrEl = card.querySelector('[class*="address"], [class*="location"]');
                    const address = addrEl ? addrEl.textContent.trim() : '';
                    if (name || href) {
                        results.push({
                            name: name || 'Restaurant ' + (i+1),
                            address: address,
                            foodpanda: href
                        });
                    }
                });
                return results;
            }
        """)

        browser.close()

    seen = set()
    unique = []
    for r in restaurants:
        k = r["name"] + r.get("foodpanda", "")
        if k not in seen:
            seen.add(k)
            unique.append(r)

    output = {
        "scrapedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "source": URL,
        "count": len(unique),
        "restaurants": unique,
    }
    OUTPUT.write_text(json.dumps(output, ensure_ascii=False, indent=2))
    print(f"爬取到 {len(unique)} 間餐廳，已儲存至 {OUTPUT}")
    return unique


if __name__ == "__main__":
    scrape()

File: scraper/scrape-foodpanda.js
/**
 * Foodpanda 香港健康餐餐廳爬蟲
 * 目標：https://www.foodpanda.hk/city/hong-kong/cuisine/healthy
 *
 * 使用方式：
 * 1. cd scraper && npm install
 * 2. npx playwright install chromium
 * 3. node scrape-foodpanda.js
 */

import { chromium } from 'playwright';
import { writeFileSync } from 'fs';

const DELAY_MS = 2000; // 每次操作間隔，避免被封
const OUTPUT_FILE = '../scraped-foodpanda.json';

async function scrapeFoodpanda() {
  console.log('啟動瀏覽器...');
  const browser = await chromium.launch({ headless: true });

  try {
    const page = await browser.newPage();

    // 模擬真實用戶
    await page.setViewportSize({ width: 1280, height: 800 });
    await page.setExtraHTTPHeaders({
      'Accept-Language': 'zh-HK,zh;q=0.9,en;q=0.8',
    });

    console.log('前往 Foodpanda 健康餐分類...');
    await page.goto('https://www.foodpanda.hk/city/hong-kong/cuisine/healthy', {
      waitUntil: 'networkidle',
      timeout: 30000,
    });

    await sleep(DELAY_MS);

    // 滾動頁面觸發懶加載（若有）
    await autoScroll(page);

    // 嘗試多種可能的選擇器（Foodpanda 結構可能變動）
    const restaurants = await page.evaluate(() => {
      const results = [];

      // 方式 1：常見的餐廳卡片 class
      const cards =
        document.querySelectorAll('[data-testid="restaurant-card"]') ||
        document.querySelectorAll('a[href*="/restaurant/"]') ||
        document.querySelectorAll('.vendor-list .vendor') ||
        document.querySelectorAll('[class*="RestaurantCard"]') ||
        document.querySelectorAll('article');

      cards.forEach((card, i) => {
        const link = card.querySelector('a[href*="/restaurant/"]') || card.closest('a');
        const href = link?.href || card.href || '';

        // 提取餐廳 ID（用於組裝 Foodpanda 連結）
        const match = href.match(/\/restaurant\/([^/]+)\/([^/?]+)/);
        const restaurantId = match ? match[1] : '';
        const slug = match ? match[2] : '';

        const nameEl =
          card.querySelector('[class*="name"]') ||
          card.querySelector('h2') ||
          card.querySelector('h3') ||
          card.querySelector('span[class*="title"]');
        const name = nameEl?.textContent?.trim() || '';

        const addressEl = card.querySelector('[class*="address"]') || card.querySelector('[class*="location"]');
        const address = addressEl?.textContent?.trim() || '';

        if (name || href) {
          results.push({
            name: name || `Restaurant ${i + 1}`,
            address,
            foodpanda: href || `https://www.foodpanda.hk/restaurant/${restaurantId}/${slug}`,
            restaurantId,
          });
        }
      });

      return results;
    });

    // 去重
    const seen = new Set();
    const unique = restaurants.filter((r) => {
      const key = r.name + r.foodpanda;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    });

    console.log(`爬取到 ${unique.length} 間餐廳`);

    const output = {
      scrapedAt: new Date().toISOString(),
      source: 'https://www.foodpanda.hk/city/hong-kong/cuisine/healthy',
      count: unique.length,
      restaurants: unique,
    };

    writeFileSync(OUTPUT_FILE, JSON.stringify(output, null, 2), 'utf-8');
    console.log(`已儲存至 ${OUTPUT_FILE}`);
    console.log('請手動核對並合併到 data.js');

    return unique;
  } finally {
    await browser.close();
  }
}

async function autoScroll(page) {
  await page.evaluate(async () => {
    await new Promise((resolve) => {
      let totalHeight = 0;
      const distance = 300;
      const timer = setInterval(() => {
        window.scrollBy(0, distance);
        totalHeight += distance;
        if (totalHeight >= document.body.scrollHeight) {
          clearInterval(timer);
          resolve();
        }
      }, 100);
    });
  });
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

scrapeFoodpanda().catch((err) => {
  console.error('爬取失敗:', err);
  process.exit(1);
});
